internal/blobstore/multipart.go

```diff
@@ -0,0 +1,226 @@
+		// knows to delete the part.
+		if err := initializePart(uploadc, uploadId, part, hash, size); err != nil {
+			return errgo.Mask(err)
+		}
+	}
+	// The part record has been updated successfully, so
+	// we can actually upload the part now.
+	partName := s.partName(uploadId, part)
+	if err := s.Put(r, partName, size, hash); err != nil {
+		return errgo.Notef(err, "cannot upload part %q", partName)
+	}
+
+	// We've put the part document, so we can
+	// now mark the part as complete.
+	err := uploadc.UpdateId(uploadId, bson.D{{
+		"$set", bson.D{{
+			partElem + ".complete", true,
+		}},
+	}})
+	if err != nil {
+		return errgo.Notef(err, "cannot mark part as complete")
+	}
+	return nil
+}
+
+// ListParts returns all the parts associated with the given
+// upload id. It returns ErrNotFound if the upload has been
+// deleted or finished.
+func (s *Store) ListParts(uploadId string) ([]Part, error) {
+	return nil, errgo.New(" not implemented yet")
+	// TODO
+	// read multipart metadata
+	// return parts from that, omitting parts that are currently in progress
+}
+
+// partName returns the blob name of the part with the given
+// uploadId and part number.
+func (s *Store) partName(uploadId string, part int) string {
+	return fmt.Sprintf("%s/%d", uploadId, part)
+}
+
+// initializePart creates the initial record for a part.
+func initializePart(uploadc *mgo.Collection, uploadId string, part int, hash string, size int64) error {
+	partElem := fmt.Sprintf("parts.%d", part)
+	err := uploadc.Update(bson.D{
+		{"_id", uploadId},
+		{"$or", []bson.D{{{
+			partElem, bson.D{{"$exists", false}},
+		}}, {{
+			partElem, bson.D{{"$eq", nil}},
+		}}}},
+	},
+		bson.D{{
+			"$set", bson.D{{partElem, uploadPart{
+				Hash: hash,
+				Size: size,
+			}}},
+		}},
+	)
+	if err == nil || err != mgo.ErrNotFound {
+		return nil
+	}
+	return errgo.New("cannot update initial part record - concurrent upload of the same part?")
+}
+
+// FinishParts completes a multipart upload by joining all the given
+// parts into one blob. The resulting blob can be opened by passing
+// uploadId and the returned multipart index to Open.
+//
+// The part numbers used will be from 0 to len(parts)-1.
+//
+// This does not delete the multipart metadata, which should still be
+// deleted explicitly by calling DeleteParts after the index data is
+// stored.
+func (s *Store) FinishParts(uploadId string, parts []Part) (idx *MultipartIndex, hash string, err error) {
+	// TODO read metadata
+	// if parts don't match uploaded parts, return error.
+	// read all parts in sequence to hash them.
+	// return index derived from metadata and calculated hash.
+	return nil, "", errgo.New("not implemented yet")
+}
+
+// DeleteExpiredParts deletes any multipart entries
+// that have passed their expiry date.
+func (s *Store) DeleteExpiredParts() error {
+	return errgo.New("not implemented yet")
+}
+
+// DeleteParts deletes all the parts associated with the
+// given upload id. It does nothing if FinishParts
```

Comment:
This seems to contradict the docstring of FinishParts, which states "This does not delete the multipart metadata, which should still be deleted explicitly by calling DeleteParts ...".